### PR TITLE
Add support for Gateway Listener certificateRefs

### LINKED ISSUE #2474: Implement Gateway certificateRef
Unlike Ingress, certificates for Gateway API Routes are [loaded via Gateway configuration](https://gateway-api.sigs.k8s.io/geps/gep-746/?h=certificate#attaching-tls-certificates-with-routes-is-confusing), not configuration on individual routes. Our current Gateway implementation does not handle the CertificateRefs field and we do not load certificates for Gateway API Routes.

The spec only provides references to TLS Secrets, which do not include any dedicated field we can populate Kong SNIs from. We need to either:
- Parse the certificate and add all hostnames present on it as SNIs.
- Add an SNIs annotation for Secrets.

These are compatible with one another. The annotation can serve as an override. I recommend just using all hostnames pending demand for the ability to omit some hostnames that a certificate is valid for or serve it for additional hostnames that it is not valid for. I do not expect any common use case to require changing the set of matched hostnames.

### PR DESCRIPTION
**Pull request by rainest**
**What this PR does / why we need it**:
- Adds store for Gateway resources and adds/removes Gateways from the store in the Gateway controller.
- Adds a new parser function to retrieve Gateway certificates, check ReferencePolicy permissions, and load them into the Kong configuration.
- Logs an error when different certificates request the same SNI.
- Updates TLSRoute tests to set certificates in Listeners and verify that expected certs are served.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #2474

**Special notes for your reviewer**:

Additional changelog item because I forgot to do one for #2555 

Appears ready, though I've seen what appear to be a few flakes in previous test runs (TestTCPIngressEssentials failed at tcpingress_test.go:124 on Enterprise integration only, where it should behave the same on all versions and should not be affected by anything in this).

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/dataplane/parser/parser.go
Comment (pmalek) [inline internal/dataplane/parser/parser.go:412]: I reckon that might be a lot but can we refactor this (whole block) so that the happy paths are the least indented ones?

e.g. instead of 

```go
			if ready {
				// ...
			}
```

->

```go
			if !ready {
				continue // or use a named continue/break if in deeply nested code
			}
```

+ extract some code blocks from this loop to named `func`s so that it's easier to read? Personally I find this a bit hard to reason about.

🥇 for nice debug logs though

File: internal/store/fake_store_test.go
Comment (pmalek) [inline internal/store/fake_store_test.go:862]:
```suggestion
	require.Nil(t, err)
	require.NotNil(t, store)
```

Without this we'll continue on with `err != nil` or `store` that is `nil` (which could panic given the below invocation).

Comment (shaneutt) [reply internal/store/fake_store_test.go:862]: ++, require whenever the rest of the test couldn't possibly meaningfully continue.

File: test/integration/tlsroute_test.go
Comment (shaneutt) [inline test/integration/tlsroute_test.go:472]: I appreciate the helpful note on this test :+1:

File: test/integration/tlsroute_test.go
Comment (shaneutt) [inline test/integration/tlsroute_test.go:556]: Similar to a previous comment, with KTF's new `cluster.Cleaner` this whole block could become:

```golang
	cleaner.Add(secret1)
	cleaner.Add(secret2)
```

File: internal/dataplane/parser/parser.go
Comment (rainest) [reply internal/dataplane/parser/parser.go:106]: It's basically if you have a TLS Ingress with `host: collision.example.com` and then create an HTTPS Listener with `hostname: collision.example.com`, the certificate on the Ingress wins. This ordering is basically arbitrary--someone has to win, but there's no rule that actually says who that is. If anything, Ingress is the established, stable implementation, so it's the one that wins, but that's a weak justification.

I think in practice I'd just recommend not using a mixture of Gateway and Ingress if possible, though that doesn't really save you since each can cause collisions on its own. Ingress has no protection against using the same host in different Ingresses (and we already did this sort of collision resolution, albeit silently for some reason) and Gateway only protects against it if you don't collapse Gateway resources (we do).

File: internal/dataplane/parser/parser.go
Comment (rainest) [reply internal/dataplane/parser/parser.go:519]: It was garbage I missed removing after refactoring.

File: CHANGELOG.md
Comment (rainest) [reply CHANGELOG.md:68]: Do we have anywhere other than that and the examples (which would break their tests now without the appropriate port)?

File: internal/controllers/gateway/gateway_utils.go
Comment (rainest) [reply internal/controllers/gateway/gateway_utils.go]: 66bf2640 rewords and moves this (it wasn't above the right check originally).

File: internal/dataplane/parser/parser.go
Comment (rainest) [reply internal/dataplane/parser/parser.go:432]: Renamed things. For the loop, I think we should avoid various single-purpose single-type functions and instead generically handle finding a Condition with certain qualities, since we do that everywhere. Are you okay deferring that to https://github.com/Kong/kubernetes-ingress-controller/issues/2609 ?

File: internal/dataplane/parser/parser.go
Comment (rainest) [reply internal/dataplane/parser/parser.go:515]: Aside from the Condition check, what are you looking to break out here? As I read this, the actual actions themselves aren't particularly complex, it's just longish because we have a bunch of multi-line struct definitions, mostly for the log fields.

File: CHANGELOG.md
Comment (shaneutt) [reply CHANGELOG.md:68]: I checked around, and it looks like the answer is no.

File: internal/dataplane/parser/parser.go
Comment (shaneutt) [reply internal/dataplane/parser/parser.go:432]: > Renamed things. For the loop, I think we should avoid various single-purpose single-type functions and instead generically handle finding a Condition with certain qualities

Can you help me to understand better why you feel we should avoid this? I don't fundamentally think there's any problem with single-purpose/single-type functions and in fact I think they can significantly improve composibility, readability, and testability of code.

> since we do that everywhere.

I guess what I'm advocating for then is to change the style here where we have a chance, rather than continuing to commit to an existing style that I think isn't as optimal.

File: internal/dataplane/parser/parser.go
Comment (shaneutt) [reply internal/dataplane/parser/parser.go:515]: I suppose what I was thinking about most was seeing if we could tighten up those log lines yes. However, this doesn't need to be a blocker.

File: test/integration/tlsroute_test.go
Comment (shaneutt) [reply test/integration/tlsroute_test.go:468]: > Verify in the TLS chain valid sense? We don't care if a cert is trusted given a set of CAs

No sorry, not this.

> just that Kong serves the certificate that we want it to

This. Did I miss it are we already doing this somewhere in here?

File: internal/dataplane/parser/parser.go
Comment (rainest) [reply internal/dataplane/parser/parser.go:405]: I don't think this should definitely change (there are reasons to prefer best effort and `parser.Build()` doesn't have final say on whether config can be applied; Kong can still reject it), but would argue that the scope here should be to make something consistent with the behavior of other parser components (which I think it does): if we want to refactor parser behavior more broadly, that should be a separate spike and discussion.

I agree that this is unlikely; it's a case of potentially being able to hit an error (here because client-go can potentially return one--it looks like only if the cached resource is not a proper Kubernetes object?) and needing to do something with it in a function that we don't want to return an error itself.

`parser.Build()` isn't consistent about using functions that log individual failures and always return successfully (possibly with an empty or partially full set), though that is the case for the majority (only `ingressRules.populateServices()` and `p.storer.ListCACerts()` errors will bubble up and break the build). `getCerts()` did skip individual certificates if it failed to load them, and `getCACerts()` just happens to have an additional conditions where it cannot return anything because it's populating from a top level resource (Gateways) rather than using outputs from another function (`getCerts()` receives a list of certificates from `ingressRulesFromIngressV1()`). The other top level resource queries do also return empty sets if their list fails.

IMO this best effort approach isn't unreasonable: if you have 99 good plugins (or consumers) and one that fails to load because it doesn't exist (a case where we currently log and omit it from the returned set), it does make sense to return those that are valid and proceed with those rather than breaking the build. This is particularly true in shared environments (common): for example, if the controller pulls from both `namespace-team-turtles` and `namespace-team-tortoises`, the turtles and tortoises teams can only access their respective namespaces, and the turtles helpfully typos their plugin name before collectively going on vacation, they'll block changes from either team and cause the tortoises considerable anguish and pain.

The cases that do break the build are also dubious: `ListCACerts()` could, like other top level resources, use a wrapper that returns an empty set (this will either break on apply because plugins reference unknown CA certs or fail safe at runtime because it cannot verify a client cert) and `populateServices()` [should only invalidate individual broken routes](https://github.com/Kong/kubernetes-ingress-controller/issues/2565). Invalid configuration de facto fails anyway down the road when Kong rejects it, and Kong is a better judge of that validity than we are.

File: test/integration/tlsroute_test.go
Comment (rainest) [reply test/integration/tlsroute_test.go:427]: If Kong did not load some certificate for that hostname, it will serve its default `*  subject: C=US; ST=California; L=San Francisco; O=Kong; OU=IT Department; CN=localhost` certificate. The CN check here confirms we're instead getting a certificate with the hostname we expect. This PR includes changes to `tlsEchoResponds()` to check the second hostname argument against the CN.

This isn't 100% foolproof because you could technically get some other certificate with `CN = tlsRouteHostname`, we just currently rely on the test environment conditions to not create that scenario

File: CHANGELOG.md
Comment (rainest) [reply CHANGELOG.md:68]: We should be set then, the resources in https://docs.konghq.com/kubernetes-ingress-controller/latest/guides/using-gateway-api/#add-a-gatewayclass-and-gateway more or less align with the tested [example](https://github.com/Kong/kubernetes-ingress-controller/blob/9dfc16a32547290ed1a990d6b336f7f11f52d968/examples/gateway-httproute.yaml#L80-L96) (it doesn't include an HTTPS port, but the guide only declares one and doesn't use it).

This will affect future guides, so https://github.com/Kong/docs.konghq.com/issues/4054

File: test/integration/tlsroute_test.go
Comment (shaneutt) [reply test/integration/tlsroute_test.go:427]: :+1:

File: internal/dataplane/parser/parser.go
Comment (rainest) [reply internal/dataplane/parser/parser.go]: Do the additional comments and newlines in 5f326535c54300407b6bf48796ec50bc921fe4f0 make this more readable? Same as https://github.com/Kong/kubernetes-ingress-controller/pull/2580/#discussion_r906293283 I don't think the internals here are that unrelated or even that complex, they just consume lots of lines because of all the debug fields an structs. I think splitting it into functions wouldn't really simplify it, since it'd be effectively the same length, but requires you to jump around to understand the full linear path.

File: internal/dataplane/parser/parser.go
Comment (shaneutt) [reply internal/dataplane/parser/parser.go]: I take it that your concern is if we do it here, then when we tried to do the bigger refactor we would somehow miss it. I feel uncertain about that stance, but at the same time this is not something that I feel needs to block this PR so in the spirit of keeping things moving and because we have a follow up issue now to look at it more holistically, let's defer to #2609 :+1:

File: internal/dataplane/parser/parser.go
Comment (shaneutt) [reply internal/dataplane/parser/parser.go:405]: I appreciate your thoughtful response. There's definitely some truth to the argument that we shouldn't be blocking a bunch of valid configs for one problem, but I think the fact that the problem exists at all is sort of intrinsically part of a greater problem. At the end of the day however this PR didn't invent this wheel so it's not responsible for the wheel unless it broke it, which it didn't. So for now I'll consider this resolved as I don't intend it to be a blocker and I'll think on this some more for later.

**Review by shaneutt — APPROVED** [review]

**Pull request merged**
